### PR TITLE
make Key an invariant functor

### PR DESCRIPTION
**Pull request by tpolecat**
Make `Key` an invariant functor, transforming input in both directions in `Locker`. See related discussion in #257 ... this can go before or after I guess, I just wanted to demonstrate what I was talking about.

### CODE REVIEW

**Comment by tpolecat** [comment]
Why not just have an `InvariantCoyoneda` as a member and delegate to it? Good question. It doesn't exist in Cats, that's wh, but my stack will overflow if I have to go further upstream. If we do `InsertKey` and `LookupKey` they can delegate to `ContravariantCoyoneda[? => A` and `Coyoneda[A => ?` and save a bit of typing.

**Comment by tpolecat** [comment]
I see that it breaks bincompat. Kind of doubt we can avoid that.

**Comment by armanbilge** [comment]
Why did Key have to become a class?

**Comment by tpolecat** [comment]
It had to become a trait because it has an abstract type.

**Comment by tpolecat** [comment]
I guess I could hide that in a member of a new data type. I'll try that.

**Comment by armanbilge** [comment]
Do abstract classes not support that? I think they do.

**Comment by tpolecat** [comment]
Ok it works now.

**Comment by tpolecat** [comment]
This capability allows a fix for https://github.com/http4s/http4s/issues/5108 which I have confirmed does work. So it would be good to get this in if we can.

**Comment by armanbilge** [comment]
Very nice!! I'm curious to see how you did it.

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,6 @@ tags
 .vscode
 .bsp
 .jekyll-cache/
+
+# metals
+metals.sbt
```
Comment (rossabaker) [inline .gitignore:16]: Protip: global gitignore. Never add this again!

**Review by rossabaker — APPROVED** [review]
:+1: This looks like a missing piece in the bug that summoned me here.  But I'd like to see that #257 can be done bincompatibly alongside this before it gets released.

**Comment by armanbilge** [comment]
@rossabaker should I base my PR on this?

**Comment by tpolecat** [comment]
@armanbilge what if I pull your changes into this branch, to see if I can answer @rossabaker's question about bincompat (it seems like it should work but I have no idea what I'm doing).

**Comment by armanbilge** [comment]
@tpolecat Thanks, that works for me!

**Comment by rossabaker** [comment]
Yes, I expect I'll :+1: #257 if it plays nicely with this.

Also, this is one of those awful two-branch repos.  I'll take it on myself to backport this if I care enough to backport the websocket fix in http4s.  I probably do.  ClassCastExceptions are bad.

**Comment by rossabaker** [comment]
#267

**Pull request closed without merging**